### ISSUE TITLE
consider equivalents to `ruff` F524, F502 and F503 rules to issue an error

### ISSUE THREAD

**Issue by andythomas** [astral-sh/ty#3542]
### Question

I am not sure where the line is drawn between `ruff` and `ty`. But the following code straight from the `ruff` rules explanations lead to a (I believe certain) runtime error. 

```python3 
# F524
"{greeting}, {name}".format(name="World")

# F502
"%(greeting)s, %(name)s" % ("Hello", "World")

# F503
"%s, %s" % {"greeting": "Hello", "name": "World"}
```

Conversely, all these rule examples are (and I believe should be) issues in both (also basically straight from the docs):

```python3
# F821
def double():
    return n * 2  # raises `NameError` if `n` is undefined when `double` is called

# F822
from pathlib import jambo

# F823
x = 1

def foo():
    x += 1


# F706
class Foo:
    return 1

# F704
class Foo:
    yield 1

# F702
def foo():
    continue  # SyntaxError

# F701
def foo():
    break

```
https://play.ty.dev/e2bae795-ff85-4ff2-820e-c7d20c5fcf9c

So my question is: Should `ty` also issue an error?

---

Background: We have an applications that provides a code editor for little measurement scripts. The user should format/use code to their liking, but should be warned about future runtime errors before running the script. We replaced `ruff` with `ty`, but these were the 3 things that are not covered. Admittedly, these are not very likely to be used.

### Version

ty 0.0.39 (020512517 2026-05-22)

**Comment by MichaReiser** [comment]
For now, no. While Ruff and ty are very similar, they do have distinct goals. 

* ty: Use the inferred and annotated types to check your program for errors. These checks have a very high certainty and tend to be less opinionated.
* Ruff: More general static analysis has a much lower bar for how accurate a rule must be. Rules are also allowed to be very opinionated

Now, the rule you suggest is interesting because it has very high accuracy (could fit into ty) but it isn't based on types (which is why it is in Ruff).

Long-term, I think this rule should be covered by ty (or should Ruff use ty?). But for now, I'd say this is a more general lint and should therefore be in Ruff, not ty. 

The other reason is that maintaining the same rule both in ty and Ruff is not free, and it can also be confusing for users to see the same error twice if they use Ruff and ty alongside. That's why we try to reduce overlap.

**Comment by andythomas** [comment]
Thank you for consideration and nice explanation. I will close.
